### DISCUSSION TITLE
.ppignore issue with folders

### DISCUSSION THREAD

**Discussion by rkmfl** [photoprism/photoprism#4349] [Q&A] [answered]
I've created a ```.ppignore``` file in the root of my ```originals``` folder, I'm running the latest version of PP 240531-60b3a4628.

Excluding files works fine; however, I cannot get folders to work.

I have searched here and googled, and I'm aware that previously indexed files will not be excluded. That's not my issue.

Here's an example of my ```.ppignore``` file:
```
@eadir
/SpotlightWallpaper99/
/Photo2/!Photography99/
/Photo2/Wallpaper99/

*-topaz.*

*.webm
*.mkv
*.flv
*.vob
*.avi
*.mov
*.wmv
*.yuv
*.mp4
*.m4v
*.mpg
*.mpeg
*.mpv
*.m2v
*.3gp
```

I've also listed the folders without the leading forward-slash:
```
SpotlightWallpaper99/
Photo2/!Photography99/
Photo2/Wallpaper99/
```

In each event I add an index number at the end of the folder name (and alter the ignore, then restart PP).

And in each of these events, it indexes that new folder that is listed in the ignore file.

Does anyone have suggestions for diagnosing this?

Thank you in advance.

**Comment by lastzero** [comment]
Did you try the folder names without slashes at the beginning and/or end?

**Reply by rkmfl** [reply]
Thank you.  Yes, I explicitly mentioned that I tested both and included both examples.

**Reply by lastzero** [reply]
All of your examples show a slash / at the end of the folder names? 🤔

**Reply by rkmfl** [reply]
Thank you.

There are two examples. One has slashes in the beginning.

The **second** example in code-brackets, shows them **without the leading slash.**

You're right, I'm missing the third example without the trailing slash; however, to your question, yes, I've tried that as well.

**Reply by lastzero** [reply]
It's the first time someone is reporting this. @graciousgrey should be able to test it with your examples next week.

PS: Note that specifying sub-directories like `Photo2/!Photography99` may not be supported, but  `SpotlightWallpaper99` should certainly work.

**Reply by rkmfl** [reply]
Yes, agreed. This is why I assumed I must be doing something wrong and posted in the Q&A (not the bug reports).

It may be some combination of parameters in my instance causes a problem.

Or as you say, it may be that sub-directories are not supported (or they require an escape for the slash, etc).

If we learn that the sub-directories are not supported, this will be great info. Either to add the limitation to the documentation, or to fix. Either/or.

Thank you.

**Reply by rkmfl** [reply, marked as answer]
I was doing more testing and you're right about the stripped directory names.

I tested without the preceding slash (indicating root) and without the trailing (indicating folder), each one removed.

However, what I did not try was removing both slashes.  In short, your original suspicion is correct: folders cannot contain leading OR trailing slashes; however, they can contain the slashes that divide folders.

Therefore, this is all valid:
```
foldername
folder1/folder2
```

Yet these are **invalid** (regardless of whether they evaluate to the correct path)
```
/foldername
foldername/
/folder1/folder2/
```

Thank you for your help.

**Reply by lastzero** [reply]
Thanks for testing it again! Exactly what I thought. If it helps, we could automatically trim slashes at the beginning and at the end?

**Reply by lastzero** [reply]
Our next release will automatically trim leading and trailing slashes from `.ppignore` patterns, see https://github.com/photoprism/photoprism/commit/4c07ec4df201420b2c58e00895d55e4e5b663526.

You can test this with our updated preview build:
- https://docs.photoprism.app/release-notes/#development-preview

**Reply by rkmfl** [reply]
@lastzero - Thank you, that's great!

Quick question. To clarify, I'm not asking for a feature; my needs are addressed. I'm asking only for sake of the community/benefitting the project...

If you **do not** use slashes to indicate directories (in a glob-formatted config file), how do you know whether to exclude a folder or a file, based on the user's input?

For example, if there is a file named:
```/MyFile```

And a folder named:
```/Folder1/MyFile/etc```

The current approach would exclude both; it appears there may not be a method to distinguish.

This is the reason I incorrectly assumed there had to be a leading or trailing space.

Again, if the preference is not to allow the distinction, that's fine, I'm not asking for anything new. I only mention in the event it's of help.

Thanks again for your patient, helpful assist!

**Reply by lastzero** [reply]
Files without extension are skipped by default, so the only possible confusion could be directories that match a filename pattern. From what I know, that's a potential issue with other ignore file implementations as well!?

**Reply by rkmfl** [reply]
> that's a potential issue with other ignore file implementations as well!?

To clarify again, the reason I used slashes is because it typically limits to a specific location (or to a folder, not file).

In typical glob-formatted file matching, these are different:
```
/ThisMustMatchRoot/**            # Matches the root -folder- explicitly
/ThisMustMatchRoot/Folder2/      # Matches an explicit -folder-
/ThisMustMatchRoot/File2         # By standard, this should only apply to a filename
**/AnyFolderName/                # Match must be Folder
**/AnyFilename                   # Match must be File
```

If the goal was the greatest compliance and versatility, this would be the standard.

With that said, I don't need these features personally. This is purely an academic conversation. If you disagree, that's certainly fine by me :)

**Reply by lastzero** [reply]
The slashes and complete glob patterns are automatically generated, so there's technically no need to add them manually and requiring to do so would break backwards compatibility for the sake of personal preferences. With the changes now implemented both works, which I believe is best as not all of our users have a background in software development or system  administration.

**Reply by rkmfl** [reply]
I agree with the backward compatibility issue, absolutely.

However, I don't think it's true that there is "no need to add them manually". The point of my example file was  to show the basis for them.

Despite the theoretical need, preserving compatibility is far more important and I think it's a non-issue.

It's still good we had the discussion as someone else will eventually face similar issues and questions and will find themselves in this thread :)
